### PR TITLE
Allow fusing epilogues whose operands are broadcast of scalar instructions.

### PR DESCRIPTION
**Pull request by elfiegg**
Allow fusing epilogues whose operands are broadcast of scalar instructions. This enables creating fusions for fp8 where the pattern is `mul(dot, scalar_ops)` where scalar ops's shapes are scalar.  This only affects epilogues, the operands of broadcast will still follow the existing fusing rules. Both triton and cuDNN backends support this kind of fusion. 

cc @sergachev

### CODE REVIEW

**Comment by NaiyerRizz** [comment]
Hi @elfiegg 
Can you please look into as few checks are failing.
Thank you

**Comment by elfiegg** [comment]
Done. Seems like a checker that disallows implicit priority between && and || .

> Hi @elfiegg Can you please look into as few checks are failing. Thank you

File: xla/service/gpu/gemm_fusion_test.cc [hunk 1]
```diff
@@ -1215,6 +1215,33 @@ ENTRY e {
...
+  p2 = f16[1] parameter(2)
+  p3 = f16[1] parameter(3)
+  m0 = f16[1] multiply(f16[1] p2, f16[1] p3)
+  bc = f16[] bitcast(m0)
+  b = f16[18,256] broadcast(f16[] bc)
+  ROOT m = f16[18,256] multiply(d, b), metadata={op_name="foo"}
+})")
+                    .value();
+  EXPECT_TRUE(GemmFusion(gpu_version_).Run(module.get()).value());
+  EXPECT_EQ(module->entry_computation()
```
Comment (sergachev) [inline xla/service/gpu/gemm_fusion_test.cc:1236]: It would be more precise to test that all instructions get fused and nothing is left outside.

Comment (elfiegg) [reply xla/service/gpu/gemm_fusion_test.cc]: That's true. Changed to matching the fusion instead.

**Review by sergachev — APPROVED** [review]

**Comment by elfiegg** [comment]
AMD ROCm fails due to merge conflict:

> Branch not suitable for integration as it does not merge cleanly: Command "git merge --ff 50491a60625189d137a1a0c4c7263487d0986ce4" returned status code 128:
> 
> stdout: 
> 
> stderr: Committer identity unknown

Any suggestions on how to proceed?

**Comment by sergachev** [comment]
> AMD ROCm fails due to merge conflict:
> Any suggestions on how to proceed?

It fails for other PRs too, for example https://github.com/openxla/xla/pull/15930, and is non-blocking.

**Comment by elfiegg** [comment]
Gentle ping @bchetioui .

File: xla/service/gpu/triton_tiling_propagation.cc [hunk 2]
```diff
@@ -1079,10 +1079,10 @@ GetPropagatedDimOrdersAndRequirementsIfProfitablyFusible(
...
         continue;
       }
-      // Currently only broadcasts of scalar constants or parameters
-      // are accepted as other inputs of non-unary operations
-      // in the output fusion.
-      if (hlo_query::IsBroadcastOfScalarConstant(*operand) ||
+      // Currently only broadcasts of scalars or parameters are accepted as
+      // other inputs of non-unary operations in the output fusion.
+      if ((operand->opcode() == HloOpcode::kBroadcast &&
+           ShapeUtil::IsEffectiveScalar(operand->operand(0)->shape())) ||
```
Comment (bchetioui) [inline xla/service/gpu/triton_tiling_propagation.cc:1085]: Are you sure that this is supported in codegen? I seemed to recall broadcast failures for non-0d shapes.

Please add a corresponding emitter test.

**Review by bchetioui — CHANGES_REQUESTED** [review]

**Comment by bchetioui** [comment]
Sorry for the delay, I missed this PR :)

File: xla/service/gpu/triton_tiling_propagation.cc (hunk 2 above)
Comment (elfiegg) [reply xla/service/gpu/triton_tiling_propagation.cc]: Good catch thanks - I tried to find some example test cases but I think there are currently only broadcast of scalars. I might have a wrong impression that we need to enable effective-scalars. 
Changed to fusing only scalar instructions. Also added test in emitter test.

File: xla/service/gpu/fusions/triton/triton_fusion_emitter_device_legacy_test.cc [hunk 3]
```diff
@@ -2077,6 +2077,29 @@ ENTRY e {
   EXPECT_TRUE(RunAndCompare(kHloText, ErrorSpec{/*aabs=*/2e-3, /*arel=*/2e-3}));
 }
 
+TEST_F(TritonGemmLevel2Test, BroadcastOfScalarIsFused) {
```
Comment (bchetioui) [inline xla/service/gpu/fusions/triton/triton_fusion_emitter_device_legacy_test.cc:2080]: I don't think `IsFused` is what we want to test here. Rather, the test is about whether the emitter works correctly---i.e. given a post-optimization Triton fusion that contains the broadcast of an effective scalar, it lowers correctly to LLVM.

I think we can rewrite this test to only create IR and compile it down to LLVM. If you think it is necessary to check the numerics (I don't think it is), then `RunAndCompareNoHloPasses` can additionally be used.

**Review by bchetioui — CHANGES_REQUESTED** [review]
Thanks! One more comment on the new test :)

Also, note that the PR/commit message seems to be out of sync, now.

File: xla/service/gpu/fusions/triton/triton_fusion_emitter_device_legacy_test.cc (hunk 3 above)
Comment (elfiegg) [reply xla/service/gpu/fusions/triton/triton_fusion_emitter_device_legacy_test.cc]: Sounds good - revised to check IR instead

Comment (bchetioui) [reply xla/service/gpu/fusions/triton/triton_fusion_emitter_device_legacy_test.cc]: Thanks, this is a bit better---but note that `CreateTritonIrAndFileCheckForDot` does not actually compile all the way down to LLVM.

You will need to call `TritonWrapper` directly to do that, as is done in [triton_support_test.cc](https://github.com/openxla/xla/blob/e46379617054479acbb2453e3871b5b0d7b9c436/xla/service/gpu/fusions/triton/triton_support_test.cc#L191) (in the absence of a better testing API).

I think calling both `CreateTritonIrAndFileCheckForDot` as you did and `TritonWrapper` is a reasonable solution for now. I'll plan to make another change allowing to both check Triton IR & compile down to LLVM in a single call.

Comment (bchetioui) [reply xla/service/gpu/fusions/triton/triton_fusion_emitter_device_legacy_test.cc]: Hi @elfiegg, any update on this?

Comment (elfiegg) [reply xla/service/gpu/fusions/triton/triton_fusion_emitter_device_legacy_test.cc]: Sorry for the delay! Done testing in TritonWrapper too.

Comment (elfiegg) [reply xla/service/gpu/fusions/triton/triton_fusion_emitter_device_legacy_test.cc]: gentle ping :)

Comment (elfiegg) [reply xla/service/gpu/fusions/triton/triton_fusion_emitter_device_legacy_test.cc]: Hi @bchetioui - friendly ping ;:^)

File: xla/service/gpu/transforms/gemm_fusion_test.cc [hunk 4]
```diff
@@ -1216,6 +1216,28 @@ ENTRY e {
...
+  p0 = f16[2,18] parameter(0)
+  p1 = f16[256,2] parameter(1)
+  d = f16[18,256] dot(p0, p1),
+    lhs_contracting_dims={0}, rhs_contracting_dims={1}
+  p2 = f16[1] parameter(2)
+  p3 = f16[1] parameter(3)
+  m0 = f16[1] multiply(f16[1] p2, f16[1] p3)
+  bc = f16[] bitcast(m0)
+  b = f16[18,256] broadcast(f16[] bc)
+  ROOT m = f16[18,256] multiply(d, b), metadata={op_name="foo"}
```
Comment (bchetioui) [inline xla/service/gpu/transforms/gemm_fusion_test.cc:1232]: This metadata seems irrelevant to the test. Let's remove it?

**Review by bchetioui — CHANGES_REQUESTED** [review]
Thanks for addressing the comments---here are a couple more, but this looks much better!

Thanks for your patience as well---I was OOO at the end of last week and just got back.

File: xla/service/gpu/fusions/triton/triton_fusion_emitter_device_legacy_test.cc [hunk 5]
```diff
@@ -2318,6 +2318,58 @@ ENTRY e {
...
+  multiply.2 = f16[] multiply(f16[] p2, f16[] p3)
+  broadcast.1 = f16[18,256]{1,0} broadcast(f16[] multiply.2), dimensions={}
+  ROOT multiply.3 = f16[18,256]{1,0} multiply(f16[18,256]{1,0} d.1, f16[18,256]{1,0} broadcast.1)
+}
+ENTRY e  {
+  p0.1 = f16[2,18]{1,0} parameter(0)
+  p1.1 = f16[256,2]{1,0} parameter(1)
+  p2.1 = f16[] parameter(2)
+  p3.1 = f16[] parameter(3)
+  ROOT %gemm_fusion_d.1 = f16[18,256]{1,0} fusion(f16[2,18]{1,0} p0.1, f16[256,2]{1,0} p1.1, f16[] p2.1, f16[] p3.1), kind=kCustom, calls=fusion, backend_config={"operation_queue_id":"0","wait_on_operation_queues":[],"fusion_backend_config":{"kind":"__triton_gemm","triton_gemm_config":{"block_m":"32","block_n":"32","block_k":"16","split_k":"1","num_stages":"1","num_warps":"4","num_ctas":"1"}},"force_earliest_schedule":false}
```
[the remark this answers is not in the document]
Comment (elfiegg) [reply xla/service/gpu/fusions/triton/triton_fusion_emitter_device_legacy_test.cc]: Done.

File: xla/service/gpu/fusions/triton/triton_fusion_emitter_device_legacy_test.cc [hunk 6]
```diff
@@ -2318,6 +2318,58 @@ ENTRY e {
   EXPECT_TRUE(RunAndCompare(kHloText, ErrorSpec{/*aabs=*/2e-3, /*arel=*/2e-3}));
 }
 
+TEST_F(TritonGemmLevel2Test, BroadcastOfScalarWorksCorrectly) {
```
[the remark this answers is not in the document]
Comment (elfiegg) [reply xla/service/gpu/fusions/triton/triton_fusion_emitter_device_legacy_test.cc]: Done. Thought broadcast needs Level2 but it seems TritonGemmTest works well.

File: xla/service/gpu/transforms/gemm_fusion_test.cc (hunk 4 above)
Comment (elfiegg) [reply xla/service/gpu/transforms/gemm_fusion_test.cc]: Done.

File: xla/service/gpu/fusions/triton/triton_fusion_emitter_device_legacy_test.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (bchetioui) [reply xla/service/gpu/fusions/triton/triton_fusion_emitter_device_legacy_test.cc]: "Level2" is relevant only when exercising the fusion logic---but this is a pure codegen test now :)

**Review by bchetioui — APPROVED** [review]
